backend-lib/src/main/rust/lib.rs

```diff
@@ -1148,23 +1016,245 @@ pub unsafe extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_re
     unwrap_exc_or(&env, res, JNI_FALSE)
 }
 
+fn decode_sapling_subtree_root(
+    env: &JNIEnv<'_>,
+    obj: JObject<'_>,
+) -> Result<CommitmentTreeRoot<sapling::Node>, failure::Error> {
+    let long_as_u32 = |name| -> Result<u32, failure::Error> {
+        Ok(u32::try_from(env.get_field(obj, name, "J")?.j()?)?)
+    };
+
+    fn byte_array(
+        env: &JNIEnv<'_>,
+        obj: JObject<'_>,
+        name: &str,
+    ) -> Result<Vec<u8>, failure::Error> {
+        let field = env.get_field(obj, name, "[B")?.l()?.into_raw();
+        Ok(env.convert_byte_array(field)?[..].try_into()?)
+    }
+
+    Ok(CommitmentTreeRoot::from_parts(
+        BlockHeight::from_u32(long_as_u32("completingBlockHeight")?),
+        sapling::Node::read(&byte_array(env, obj, "rootHash")?[..])?,
+    ))
```

Comment:
The names `long_as_u32` and `byte_array` don't imply that they do field access. Also, there seems to be no reason for the former to be a closure and the latter to be a function, or for the `[..].try_into()?` in the latter (since [JNIEnv::convert_byte_array](https://docs.rs/jni/latest/jni/struct.JNIEnv.html#method.convert_byte_array) already returns a `Result` of `Vec<u8>`).

```suggestion
    let long_field = |name| -> Result<u32, failure::Error> {
        Ok(u32::try_from(env.get_field(obj, name, "J")?.j()?)?)
    };

    let byte_array_field = |name| -> Result<Vec<u8>, failure::Error> {
        let field = env.get_field(obj, name, "[B")?.l()?.into_raw();
        Ok(env.convert_byte_array(field)?)
    }

    Ok(CommitmentTreeRoot::from_parts(
        BlockHeight::from_u32(long_field("completingBlockHeight")?),
        sapling::Node::read(&byte_array_field("rootHash")?[..])?,
    ))
```
Non-blocking.

Reply:
Note that there is similar code in `decode_blockmeta`. For consistency it can also be changed to:
```rust
fn decode_blockmeta(env: &JNIEnv<'_>, obj: JObject<'_>) -> Result<BlockMeta, failure::Error> {
    let long_field = |name| -> Result<u32, failure::Error> {
        Ok(u32::try_from(env.get_field(obj, name, "J")?.j()?)?)
    };

    let byte_array_field = |name| -> Result<Vec<u8>, failure::Error> {
        let field = env.get_field(obj, name, "[B")?.l()?.into_raw();
        Ok(env.convert_byte_array(field)?)
    };

    Ok(BlockMeta {
        height: BlockHeight::from_u32(long_field("height")?),
        block_hash: BlockHash(byte_array_field("hash")?[..].try_into()?),
        block_time: long_field("time")?,
        sapling_outputs_count: long_field("saplingOutputsCount")?,
        orchard_actions_count: long_field("orchardOutputsCount")?,
    })
}
```

Reply:
It was that the function was generic on the array length in the case of `decode_blockmeta`, and you can't have generic closures. But the code above solves that more simply by having the closure return `Result<Vec<u8>, failure::Error>`. It's no less efficient because the closure will be inlined anyway, given that it has only one use.

Reply:
Anyway this is non-blocking.